src/sqlancer/stonedb/StoneDBSchema.java

```diff
@@ -0,0 +1,393 @@
+        }
+    }
+
+    public static StoneDBSchema fromConnection(SQLConnection con, String databaseName) throws SQLException {
+        List<StoneDBTable> databaseTables = new ArrayList<>();
+        List<String> tableNames = getTableNames(con, databaseName);
+        for (String tableName : tableNames) {
+            List<StoneDBColumn> databaseColumns = getTableColumns(con, databaseName, tableName);
+            List<StoneDBIndex> indexes = getIndexes(con, databaseName, tableName);
+            boolean isView = tableName.startsWith("v");
+            StoneDBTable t = new StoneDBTable(tableName, databaseColumns, indexes, isView);
+            for (StoneDBColumn c : databaseColumns) {
+                c.setTable(t);
+            }
+            databaseTables.add(t);
+
+        }
+        return new StoneDBSchema(databaseTables);
+    }
+
+    private static List<StoneDBIndex> getIndexes(SQLConnection con, String databaseName, String tableName)
+            throws SQLException {
+        List<StoneDBIndex> indexes = new ArrayList<>();
+        try (ResultSet rs = con.createStatement()
+                .executeQuery("SELECT INDEX_NAME FROM INFORMATION_SCHEMA.STATISTICS WHERE TABLE_SCHEMA = '"
+                        + databaseName + "' AND TABLE_NAME='" + tableName + "';")) {
+            while (rs.next()) {
+                String indexName = rs.getString("INDEX_NAME");
+                indexes.add(StoneDBIndex.create(indexName));
+            }
+        }
+        return indexes;
+    }
+
+    private static List<StoneDBColumn> getTableColumns(SQLConnection con, String databaseName, String tableName)
+            throws SQLException {
+        List<StoneDBColumn> columns = new ArrayList<>();
+        try (ResultSet rs = con.createStatement()
+                .executeQuery("select * from information_schema.columns where table_schema = '" + databaseName
+                        + "' AND TABLE_NAME='" + tableName + "';")) {
+            while (rs.next()) {
+                String columnName = rs.getString("COLUMN_NAME");
+                String dataType = rs.getString("DATA_TYPE");
+                int precision = rs.getInt("NUMERIC_PRECISION");
+                boolean isNullable = !rs.getString("IS_NULLABLE").equals("NO");
+                boolean isPrimaryKey = rs.getString("COLUMN_KEY").equals("PRI");
+                StoneDBColumn c = new StoneDBColumn(columnName, getColumnCompositeDataType(dataType), isPrimaryKey,
+                        isNullable, precision);
+                columns.add(c);
+            }
+        }
+        return columns;
+    }
+
+    private static StoneDBCompositeDataType getColumnCompositeDataType(String typeString) {
+        if (Arrays.stream(StoneDBDataType.values()).noneMatch(e -> e.name().equals(typeString.toUpperCase()))) {
+            throw new AssertionError(typeString);
+        }
+        return new StoneDBCompositeDataType(StoneDBDataType.valueOf(typeString.toUpperCase()));
+    }
+
+    //
```

Comment:
I guess we could remove the commented-out code?

Reply:
will refactor this.